Initial implementation of RTCMEM storage

mcspr opened the pull request:
Part of #1265 

Store system crash counter and reset reason in rtcmem instead of eeprom
Store relay state mask in rtc in addition to the eeprom
Store relay state in eeprom only when boot mode requires it
Simplify relay state mask calculation / reading via std::bitset

Missing:
energy counter (`EEPROM_ENERGY_COUNT`) - should it be saved like this?
mqtt json id counter (`EEPROM_MESSAGE_ID`) - can write here every send, no need for delay like for eeprom.

*Note:*
This also can directly use rtcmem as an array (example [from esp-open-rtos](https://github.com/SuperHouse/esp-open-rtos/blob/a8c60e096093e9e9f4a60b885676adc2cf5b790a/core/include/esp/rtcmem_regs.h))
```c
typedef volatile uint32_t rtcmem_array_t[96];
#define RTCMEM_USER (*(rtcmem_array_t *)(RTCMEM_ADDR))
...
RTCMEM_USER[...SOME_INDEX_DEFINE...] = 1; // indexes defined like for eeprom
RTCMEM_USER[97] = 2; // compiler will yell because of rtcmem_array_t definition
```

And idk if consistency check is required at all, and just checking magic value would suffice
\+ tracking `resetInfo->reason` for `REASON_DEFAULT_RST` / `REASON_SOFT_RESTART`

xoseperez commented:
We are not checking for consistency in the flash atm.
I would add energy count too and light colors also (same update frequency than relays).

mcspr commented:
for energy counter the goal here is to protect readings in between save intervals from disappearing?

lights are straightforward enough, save 5 channels + brightness + mireds state. even fits inside 64 bit. only unlike relay it always saves - should it have an option like `relayBoot` to optionally do it?

mcspr commented:
Removed crc, added mqtt counter, added general methods for energy total and lights handling.
Last thing - check when it is appropriate to save them.

For lights: LIGHT_SAVE_ENABLED should be a setting and somehow wired into this?
For sensor: one can also save uptime() in rtc and reference that in energy ts code, since now it becomes 2 states - uptime under power with occasional reboots and complete shutdown with eeprom saves.

xoseperez commented on code/espurna/light.ino:
```diff
@@ -469,6 +469,39 @@ void _lightProviderUpdate() {
+
+void _rtcmemLightSave() {
+    rtcmem_light_t light;
+    for (unsigned int i=0; i < lightChannels(); i++) {
```
We don't support any device with more than 5 channels (are there any?) but maybe we should check here the `lightChannels()` return value.

mcspr replied:
Yeah, that is true.
Unlike pwm channels init this cannot really depend on LIGHT_CHANNELS to init array size. `rtcmem_light_t::channels` size is still hardcoded, but this can be `std::min(lightChannels(), sizeof(rtcmem_light_t::channels))` instead of simply using it.

mcspr commented:
Light issue resolved via static check of array size.

To reiterate - breaking changes are:
mqtt counter (no longer persisted between power-offs)
emon sensor total persists between reboots

I wonder if ntp time can be saved too.

mcspr commented:
No obvious things stand up. Merging in the interest of moving milestone forward.

Exception system might need some tweaking though:
- maybe lock the boards in safe mode until reboot. don't reset system stability counter to 0 on the 5th time, and force manual intervention (terminal command / web switch / power cycle)
- allow to store "small" crash log, instead of the full one
